### PR TITLE
Splitting Bigtable Row into 3 classes based on the mutation type / RPC method used in commit()

### LINKED ISSUE #1548: Find a better name for Row.commit_modifications() or fold the behavior into
See discussion in #1547 for context.

The name `modifications` is intended to map to the RPC method name `ReadModifyWriteRow` but the word modification is a synonym of mutation, so this method is poorly named. The method actually sends changes that

- increment an int (stored in BT as 8 bytes)
- append bytes to a cell value

so we should either

- Fold it into `Row.commit()` as `commit(append_only=True)`
- Rename it to something like `Row.commit_cell_updates` (or a better name than that)

### PR DESCRIPTION
**Pull request by dhermes**
Docs are still being updated, but I wanted to get this out there to speed up the review. Alternate implementation based on #1550.

I strongly urge reviewers to examine this commit-by-commit. The commit messages are fairly helpful (for the most part) and the diff of the PR as a whole is not so easy on the eyes.

Fixes #1548.

/cc @jonparrott @jgeewax

### CODE REVIEW

File: gcloud/bigtable/table.py [hunk 1]
```diff
@@ -95,20 +97,38 @@ def column_family(self, column_family_id, gc_rule=None):
...
         :param filter_: (Optional) Filter to be used for conditional mutations.
-                        See :class:`.Row` for more details.
+                        See :class:`.DirectRow` for more details.
+
+        :type append: bool
+        :param append: (Optional) Flag to determine if the row should be used
+                       for append mutations.
 
-        :rtype: :class:`.Row`
+        :rtype: :class:`.DirectRow`
```
Comment (tseaver) [inline gcloud/bigtable/table.py]: `AppendRow` doesn't derive from `DirectRow`, which makes `Row` the only signular choice.  You could signal "one of `DirectRow`, `ConditionalRow`, or `AppendRow`, maybe,

Comment (dhermes) [inline gcloud/bigtable/table.py]: I think the right move would instead be to have a private class as the shared parent of `DirectRow` and `ConditionalRow` so `isinstance(cond_row, DirectRow)` doesn't succeed and confuse people. WDYT?

**Comment by dhermes** [comment]
@tseaver just saw tests were failing on Travis. Looking into it.

What do we need to do to keep this review moving? Really close to the finish line and want to get a release out

**Comment by dhermes** [comment]
OK the failure is simple to address (just a long failure)

**Comment by dhermes** [comment]
*lint failure. Phone autocorrect!

File: gcloud/bigtable/table.py (hunk 1 above)
Comment (jonparrott) [inline gcloud/bigtable/table.py]: Sounds reasonable.

**Comment by jonparrott** [comment]
This mostly LGTM, pending the one open item.

File: gcloud/bigtable/table.py (hunk 1 above)
Comment (tseaver) [inline gcloud/bigtable/table.py]: SGTM

**Comment by tseaver** [comment]
LGTM

**Comment by dhermes** [comment]
Merging this, will handle the open item in a small separate PR.

**Pull request merged**